pkg/cloudprovider/providers/gce/gce_loadbalancer_external_test.go

```diff
@@ -953,3 +953,98 @@ func TestEnsureExternalLoadBalancerDeletedSucceedsOnXPN(t *testing.T) {
+		fakeApiService.DeepCopy(),
+		nil,
+		nodes,
+	}
+}
+
+// callEnsureExternalLoadBalancerAndCheckError wraps calling for
+// ensureExternalLoadBalancer, so TestEnsureExternalLoadBalancerErrors can be
+// easily maintained if the interface changes.
+func callEnsureExternalLoadBalancerAndCheckError(t *testing.T, gce *GCECloud, params *EnsureELBParams) {
+	status, err := gce.ensureExternalLoadBalancer(
+		params.clusterName,
+		params.clusterID,
+		params.apiService,
+		params.existingFwdRule,
+		params.nodes,
+	)
+	assert.Error(t, err)
+	assert.Nil(t, status)
+}
+
+func TestEnsureExternalLoadBalancerErrors(t *testing.T) {
+	vals := DefaultTestClusterValues()
+	gce, err := fakeGCECloud(DefaultTestClusterValues())
+	nodes, err := createAndInsertNodes(gce, []string{"test-node-1"}, vals.ZoneName)
+	require.NoError(t, err)
+	c := gce.c.(*cloud.MockGCE)
+	var params *EnsureELBParams
+
+	params = newEnsureELBParams(nodes)
+	params.nodes = []*v1.Node{}
+	callEnsureExternalLoadBalancerAndCheckError(t, gce, params)
+
+	params = newEnsureELBParams(nodes)
+	params.nodes = []*v1.Node{{}}
+	callEnsureExternalLoadBalancerAndCheckError(t, gce, params)
+
+	params = newEnsureELBParams(nodes)
+	c.MockForwardingRules.GetHook = mock.GetForwardingRulesInternalErrHook
+	callEnsureExternalLoadBalancerAndCheckError(t, gce, params)
+	c.MockForwardingRules.GetHook = nil
+
+	params = newEnsureELBParams(nodes)
+	c.MockAddresses.GetHook = mock.GetAddressesInternalErrHook
+	callEnsureExternalLoadBalancerAndCheckError(t, gce, params)
+	c.MockAddresses.GetHook = nil
+
+	params = newEnsureELBParams(nodes)
+	params.apiService.Spec.LoadBalancerSourceRanges = []string{"BadSourceRange"}
+	callEnsureExternalLoadBalancerAndCheckError(t, gce, params)
+
+	params = newEnsureELBParams(nodes)
+	c.MockFirewalls.GetHook = mock.GetFirewallsUnauthorizedErrHook
+	callEnsureExternalLoadBalancerAndCheckError(t, gce, params)
+	c.MockFirewalls.GetHook = nil
+
+	params = newEnsureELBParams(nodes)
+	c.MockFirewalls.InsertHook = mock.InsertFirewallsUnauthorizedErrHook
+	callEnsureExternalLoadBalancerAndCheckError(t, gce, params)
+	c.MockFirewalls.InsertHook = nil
+
+	params = newEnsureELBParams(nodes)
+	c.MockTargetPools.GetHook = mock.GetTargetPoolInternalErrHook
+	callEnsureExternalLoadBalancerAndCheckError(t, gce, params)
+	c.MockTargetPools.GetHook = nil
+
+	params = newEnsureELBParams(nodes)
+	c.MockHttpHealthChecks.GetHook = mock.GetHTTPHealthChecksInternalErrHook
+	callEnsureExternalLoadBalancerAndCheckError(t, gce, params)
+	c.MockHttpHealthChecks.GetHook = nil
+
+	params = newEnsureELBParams(nodes)
+	c.MockTargetPools.InsertHook = mock.InsertTargetPoolsInternalErrHook
+	callEnsureExternalLoadBalancerAndCheckError(t, gce, params)
+	c.MockTargetPools.InsertHook = nil
+
+	params = newEnsureELBParams(nodes)
+	c.MockForwardingRules.InsertHook = mock.InsertForwardingRulesInternalErrHook
+	callEnsureExternalLoadBalancerAndCheckError(t, gce, params)
+	c.MockForwardingRules.InsertHook = nil
```

Comment:
Would be great to group above testcases into a for loop :)

Reply:
I tried but I can't figure out a way to unify the mocks - they are all so different.

Reply:
We can do something flexible in this case (feel free to ignore if it doesn't make sense...):
```golang
	testCases := []struct {
		desc string
		adjustParams func(*EnsureELBParams)
		injectMock func(*cloud.MockGCE)
		resetMock func(*cloud.MockGCE)
	}{...}

	for _, tc := range() {
		params = newEnsureELBParams(nodes)
		if tc.adjustParams != nil {
			tc.adjustParams(params)
		}
		if tc.injectMock != nil {
			tc.injectMock(c)
		}
		callEnsureExternalLoadBalancerAndCheckError(t, gce, params)
		if tc.resetMock != nil {
			tc.resetMock(c)
		}
	}
``